pypeit/sensfunc.py

```diff
@@ -1,164 +1,383 @@
+#                                          descr='Spliced-together sensitivity function zeropoints'),
+#                 'throughput_splice': dict(otype=np.ndarray, atype=float,
+#                                           descr='Spliced-together spectrograph throughput '
+#                                                 'measurements')}
+    """DataContainer datamodel."""
+
+    algorithm = None
+    """Algorithm used for the sensitivity calculation."""
+
+    @staticmethod
+    def empty_sensfunc_table(norders, nspec, ncoeff=1):
+        """
+        Construct an empty `astropy.table.Table`_ for the sensitivity
+        function.
 
+        Args:
+            norders (:obj:`int`):
+                The number of slits/orders on the detector.
+            nspec (:obj:`int`):
+                The number of spectral pixels on the detector.
+
+        Returns:
+            `astropy.table.Table`_: Instance of the empty sensitivity
+            function table.
+        """
+        return table.Table(data=[
+            table.Column(name='SENS_WAVE', dtype=float, length=norders, shape=(nspec,),
+                         description='Wavelength vector'),
+            table.Column(name='SENS_COUNTS_PER_ANG', dtype=float, length=norders, shape=(nspec,),
+                         description='Flux in counts per angstrom'),
+            table.Column(name='SENS_ZEROPOINT', dtype=float, length=norders, shape=(nspec,),
+                         description='Measured sensitivity zero-point data'),
+            table.Column(name='SENS_ZEROPOINT_GPM', dtype=bool, length=norders, shape=(nspec,),
+                         description='Good-pixel mask for the measured zero points'),
+            table.Column(name='SENS_ZEROPOINT_FIT', dtype=float, length=norders, shape=(nspec,),
+                         description='Best-fit smooth model to the zero points'),
+            table.Column(name='SENS_ZEROPOINT_FIT_GPM', dtype=bool, length=norders, shape=(nspec,),
+                         description='Good-pixel mask for the model zero points'),
+            table.Column(name='SENS_COEFF', dtype=float, length=norders, shape=(ncoeff,),
+                         description='Coefficients of smooth model fit to zero points'),
+            table.Column(name='ECH_ORDERS', dtype=int, length=norders,
+                         description='Echelle order for this specrum (echelle data only)'),
+            table.Column(name='POLYORDER_VEC', dtype=int, length=norders,
+                         description='Polynomial order for each slit/echelle (if applicable)'),
+            table.Column(name='WAVE_MIN', dtype=float, length=norders,
+                         description='Minimum wavelength included in the fit'),
+            table.Column(name='WAVE_MAX', dtype=float, length=norders,
+                         description='Maximum wavelength included in the fit')])
+
+    # TODO: par is *not* optional for this method!
```

Comment:
I did  disagree that par is not optional. Par is and should be optional, because we know which spectrograph this is from the object file, and can use a sensible set of defaults unless the user desires otherwise.

Reply:
My point is that the code needs `par['algorithm']` to be able to instantiate the correct subclass.  I.e.:
```
>>> from pypeit.sensfunc import SensFunc
>>> SensFunc.get_instance('spec1d_dummy.fits', 'sensdummy.fits')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/westfall/Work/packages/pypeit/pypeit/sensfunc.py", line 147, in get_instance
    return next(c for c in cls.__subclasses__()
  File "/Users/westfall/Work/packages/pypeit/pypeit/sensfunc.py", line 148, in <genexpr>
    if c.__name__ == f"{par['algorithm']}SensFunc")(spec1dfile, sensfile, par=par,
TypeError: 'NoneType' object is not subscriptable
```
If it can get the correct algorithm from the spec1d file, then we should be using that instead of `par['algorithm']`.

Reply:
Yes you are correct. I am loathe to make spectrograph a required variable when it can be easily pulled out of the spec1dfile header but it seems that either algorithm needs to be a required argument or par must be required. Otherwise we would need to pull in the spectrograph from the header, instantiate spectrograph object, and then get default pypeit_par to get the parset, and then get the algorithm. Seems silly given that we are doing exactly that in the __init__. 

Do you see any obvious way out?

Reply:
I just did the easiest thing, which was to make `par` a required argument.